pallets/liquidity-pools/src/inbound.rs

```diff
@@ -173,80 +200,165 @@ impl<T: Config> Pallet<T> {
 			false,
 		)?;
 
-		T::ForeignInvestment::update_redemption(&investor, invest_id, post_amount)?;
+		T::ForeignInvestment::increase_foreign_redemption(&investor, invest_id, amount)?;
 
 		Ok(())
 	}
 
 	/// Decreases an existing redemption order of the investor.
 	///
 	/// Initiates a return `ExecutedDecreaseRedemption` message to refund the
-	/// decreased amount on the source domain. The dispatch of this message is
-	/// delayed until the execution of the redemption, e.g. at least until the
-	/// next epoch transition.
-	pub fn handle_decrease_redemption(
+	/// decreased amount on the source domain.
+	///
+	/// NOTE: In contrast to investments, redemption decrements happen
+	/// fully synchronously as they can only be called in between increasing a
+	/// redemption and its (full) processing.
+	pub fn handle_decrease_redeem_order(
 		pool_id: T::PoolId,
 		tranche_id: T::TrancheId,
 		investor: T::AccountId,
 		currency_index: GeneralCurrencyIndexOf<T>,
 		amount: <T as Config>::Balance,
-		_sending_domain: DomainAddress,
+		destination: DomainAddress,
 	) -> DispatchResult {
-		// Retrieve investment details
 		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
-		// NOTE: Required for relaying `ExecutedDecreaseRedemption` message
-		let _currency = Self::try_get_payment_currency(invest_id.clone(), currency_index)?;
+		// TODO(@review): This is exactly `amount` as we can only decrement up to the
+		// unprocessed redemption
+		let tranche_tokens_payout = T::ForeignInvestment::decrease_foreign_redemption(
+			&investor,
+			invest_id.clone(),
+			amount,
+		)?;
 
-		// Determine post adjustment amount
-		let pre_amount = T::ForeignInvestment::redemption(&investor, invest_id.clone())?;
-		let post_amount = pre_amount.ensure_sub(amount)?;
+		T::Tokens::transfer(
+			invest_id.into(),
+			&investor,
+			&Domain::convert(destination.domain()),
+			tranche_tokens_payout,
+			false,
+		)?;
 
-		T::ForeignInvestment::update_redemption(&investor, invest_id, post_amount)?;
+		let message: MessageOf<T> = Message::ExecutedDecreaseRedeemOrder {
```

Comment:
NOTE: AFAIU, it is correct to dispatch the Executed* messages as part of `pallet-liquidity-pools` in the context of having linked pools in the future.